pdns/ueberbackend.cc

```diff
@@ -344,15 +345,31 @@ bool UeberBackend::getAuth(const DNSName &target, const QType& qtype, SOAData* s
         }
       }
 
-      // Add to cache
-      if(i == backends.end()) {
-        if(d_negcache_ttl) {
-          DLOG(g_log<<Logger::Error<<"add neg cache entry:"<<shorter<<endl);
-          d_question.qname=shorter;
-          addNegCache(d_question);
-        }
-        continue;
-      } else if(d_cache_ttl) {
+      // If all backends have return a result (which may be empty or shorter
+      // than the current question), then we can just finish with whatever is
+      // the longest sd->qname..
+      if(i == backends.end()) {     // if != then we had an exact match and breaked earlier
+          auto longest = &bestmatch[0];
+          for( j = bestmatch.begin(); j != bestmatch.end(); ++j ) {
+              if( j->first > longest->first )
+                longest = &*j;
+          }
+
+          // All backends returned a match, get the longest
+          if( longest->first < initial_length ) {
+            *sd = longest->second;
+          } else {      // Add to negcache
+            if(d_negcache_ttl) {
+              DLOG(g_log<<Logger::Error<<"add neg cache entry:"<<shorter<<endl);
+              d_question.qname=shorter;
+              addNegCache(d_question);
+            }
+            continue;
+          }
+      }
+
+      // Definately got a result from one of the backends, cache it
```

Comment:
```suggestion
      // Definitely got a result from one of the backends, cache it
```
per https://github.com/jsoref/pdns/commit/3d71a43f243eeb6b8f4361e81ae0bfe851c935e0#commitcomment-39230327